feat: Support waiting for different client states

rainerschoe opened the pull request:
closes #67 

This implements waiting functionality to allow the user to wait for the client to reach online state.

rainerschoe committed:
implement waiting for config and online (first draft)

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
fix tests

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
Handle Fallback data + change demo + manual testing

Signed-off-by: Travis <rschoe@de.ibm.com>

renovate[bot] committed:
chore(deps): update actions/setup-python action to v6 (#110)

This PR contains the following updates:

| Package | Type | Update | Change |
|---|---|---|---|
|
[actions/setup-python](https://redirect.github.com/actions/setup-python)
| action | major | `v5` -> `v6` |

---

### Release Notes

<details>
<summary>actions/setup-python (actions/setup-python)</summary>

###
[`v6`](https://redirect.github.com/actions/setup-python/compare/v5...v6)

[Compare
Source](https://redirect.github.com/actions/setup-python/compare/v5...v6)

</details>

---

### Configuration

📅 **Schedule**: Branch creation - At any time (no schedule defined),
Automerge - At any time (no schedule defined).

🚦 **Automerge**: Disabled by config. Please merge this manually once you
are satisfied.

♻ **Rebasing**: Whenever PR becomes conflicted, or you tick the
rebase/retry checkbox.

🔕 **Ignore**: Close this PR and you won't be reminded about this update
again.

---

- [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check
this box

---

This PR was generated by [Mend Renovate](https://mend.io/renovate/).
View the [repository job
log](https://developer.mend.io/github/IBM/appconfiguration-rust-sdk).

<!--renovate-debug:eyJjcmVhdGVkSW5WZXIiOiI0MS45MS4xIiwidXBkYXRlZEluVmVyIjoiNDEuOTEuMSIsInRhcmdldEJyYW5jaCI6Im1haW4iLCJsYWJlbHMiOltdfQ==-->

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
test: Ensure we use the same hash as Node client. (#100)

Refactor the hashing code a little + add test to ensure we use the
correct hashing algorithm.

closes #2

---------

Signed-off-by: Travis <rschoe@de.ibm.com>

jgsogo reviewed (commented):
I like the implementation, but I really think it's time for a new trait related to online configurations.

rainerschoe committed:
fix: Use correct metering endpoint (#114)

closes #15

also adds some logging

We ran a manual test using `demo.rs`

<img width="1107" height="311" alt="image"
src="https://github.com/user-attachments/assets/e1dbde15-ad30-42c5-852e-fdb3b34c2e3c"
/>

Now we successfully upload metering data. See screenshot above. Last
evaluated is set to a recent data. Which is the acceptance criteria :-)

---------

Signed-off-by: Travis <rschoe@de.ibm.com>

jgsogo committed:
chore: Fix detect-secrets pre-commit (#119)

https://github.com/IBM/detect-secrets/issues/173

Override boxsdk requirement

Signed-off-by: Javier G. Sogo <jgsogo@gmail.com>
Signed-off-by: Travis <rschoe@de.ibm.com>

jgsogo committed:
feat: Add new operators for segment rules (#117)

* Add `isNot`, `notContains`, `notStartsWith`, `notEndsWith` operators
 * Refactor for clarity

closes #101

---------

Signed-off-by: Javier G. Sogo <jgsogo@gmail.com>
Signed-off-by: Travis <rschoe@de.ibm.com>

renovate[bot] committed:
chore(deps): update rust crate httpmock to 0.8.0 (#121)

Coming soon: The Renovate bot (GitHub App) will be renamed to Mend. PRs
from Renovate will soon appear from 'Mend'. Learn more
[here](https://redirect.github.com/renovatebot/renovate/discussions/37842).

This PR contains the following updates:

| Package | Type | Update | Change |
|---|---|---|---|
| [httpmock](https://redirect.github.com/httpmock/httpmock) |
dev-dependencies | minor | `0.7.0` -> `0.8.0` |

---

### Release Notes

<details>
<summary>httpmock/httpmock (httpmock)</summary>

###
[`v0.8.0`](https://redirect.github.com/httpmock/httpmock/blob/HEAD/CHANGELOG.md#Version-080)

[Compare
Source](https://redirect.github.com/httpmock/httpmock/compare/v0.7.0...v0.8.0)

This release includes refactoring, dependency updates, and internal
cleanups.
No breaking changes expected.

The minimum required Rust version has been increased to 1.82.

The following pull requests have been merged:

- [#&#8203;172](https://redirect.github.com/httpmock/httpmock/pull/172):
"Update Rust edition to 2021" (thanks
[@&#8203;FalkWoldmann](https://redirect.github.com/FalkWoldmann))
- [#&#8203;169](https://redirect.github.com/httpmock/httpmock/pull/169):
"Proxy HTTPS fix"
- [#&#8203;167](https://redirect.github.com/httpmock/httpmock/pull/167):
"Replace log and env\_logger with tracing and tracing-subscriber"
(thanks
[@&#8203;FalkWoldmann](https://redirect.github.com/FalkWoldmann))
- [#&#8203;166](https://redirect.github.com/httpmock/httpmock/pull/166):
"Remove unused code" (thanks
[@&#8203;FalkWoldmann](https://redirect.github.com/FalkWoldmann))
- [#&#8203;163](https://redirect.github.com/httpmock/httpmock/pull/163):
"fix: issue 162, non localhost hosts match" (thanks
[@&#8203;Thomblin](https://redirect.github.com/Thomblin))
- [#&#8203;160](https://redirect.github.com/httpmock/httpmock/pull/160):
"Replace custom read\_file with std::fs::read\_to\_string" (thanks
[@&#8203;FalkWoldmann](https://redirect.github.com/FalkWoldmann))
- [#&#8203;158](https://redirect.github.com/httpmock/httpmock/pull/158):
"Improve async executor support"
- [#&#8203;156](https://redirect.github.com/httpmock/httpmock/pull/156):
"Bump async-object-pool to replace async-std"
- [#&#8203;153](https://redirect.github.com/httpmock/httpmock/pull/153):
"ci(deps): bump actions/checkout from 4 to 5"
- [#&#8203;152](https://redirect.github.com/httpmock/httpmock/pull/152):
"Fix missing standalone routes"
- [#&#8203;151](https://redirect.github.com/httpmock/httpmock/pull/151):
"Cleanup unused test functions"
- [#&#8203;147](https://redirect.github.com/httpmock/httpmock/pull/147):
"ci(deps): bump codecov/codecov-action from 2 to 5"
- [#&#8203;146](https://redirect.github.com/httpmock/httpmock/pull/146):
"cargo(deps): update thiserror requirement from 1 to 2"
- [#&#8203;145](https://redirect.github.com/httpmock/httpmock/pull/145):
"ci(deps): bump actions/checkout from 2 to 4"
- [#&#8203;144](https://redirect.github.com/httpmock/httpmock/pull/144):
"ci(deps): bump docker/build-push-action from 4 to 6
- [#&#8203;141](https://redirect.github.com/httpmock/httpmock/pull/141):
"cargo(deps): update path-tree requirement from >=0.8.0, <0.8.1 to
>=0.8.0, <0.8.4"
- [#&#8203;140](https://redirect.github.com/httpmock/httpmock/pull/140):
"ci(deps): bump docker/login-action from 1 to 3"
- [#&#8203;139](https://redirect.github.com/httpmock/httpmock/pull/139):
"ci(deps): bump withastro/action from 2 to 4"
- [#&#8203;138](https://redirect.github.com/httpmock/httpmock/pull/138):
"Create dependabot.yml" (thanks
[@&#8203;FalkWoldmann](https://redirect.github.com/FalkWoldmann))
- [#&#8203;136](https://redirect.github.com/httpmock/httpmock/pull/136):
"Replace async\_std with tokio" (thanks
[@&#8203;FalkWoldmann](https://redirect.github.com/FalkWoldmann))

</details>

---

### Configuration

📅 **Schedule**: Branch creation - At any time (no schedule defined),
Automerge - At any time (no schedule defined).

🚦 **Automerge**: Disabled by config. Please merge this manually once you
are satisfied.

♻ **Rebasing**: Whenever PR is behind base branch, or you tick the
rebase/retry checkbox.

🔕 **Ignore**: Close this PR and you won't be reminded about this update
again.

---

- [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check
this box

---

This PR was generated by [Mend Renovate](https://mend.io/renovate/).
View the [repository job
log](https://developer.mend.io/github/IBM/appconfiguration-rust-sdk).

<!--renovate-debug:eyJjcmVhdGVkSW5WZXIiOiI0MS4xMzAuMSIsInVwZGF0ZWRJblZlciI6IjQxLjEzMC4xIiwidGFyZ2V0QnJhbmNoIjoibWFpbiIsImxhYmVscyI6W119-->

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Signed-off-by: Travis <rschoe@de.ibm.com>

renovate[bot] committed:
fix(deps): update rust crate tungstenite to 0.28.0 (#120)

Coming soon: The Renovate bot (GitHub App) will be renamed to Mend. PRs
from Renovate will soon appear from 'Mend'. Learn more
[here](https://redirect.github.com/renovatebot/renovate/discussions/37842).

This PR contains the following updates:

| Package | Type | Update | Change |
|---|---|---|---|
| [tungstenite](https://redirect.github.com/snapview/tungstenite-rs) |
dependencies | minor | `0.27.0` -> `0.28.0` |

---

### Release Notes

<details>
<summary>snapview/tungstenite-rs (tungstenite)</summary>

###
[`v0.28.0`](https://redirect.github.com/snapview/tungstenite-rs/blob/HEAD/CHANGELOG.md#0280)

[Compare
Source](https://redirect.github.com/snapview/tungstenite-rs/compare/v0.27.0...v0.28.0)

- Reduce `Error` size 136 -> **32** by boxing internals of
`Error::Http`, `Error::WriteBufferFull`,
`ProtocolError::InvalidHeader`, `TlsError::Native`, `TlsError::Rustls`.
- Dependency update (`socket` to `0.6.0`).
- Add `into_inner()` to the `WebSocket`, so that the user can extract
the underlying stream.
- Address the edge-case for the WebSocket request generation when
`tungstenite` is built with `-Zfmt-debug=none`.

</details>

---

### Configuration

📅 **Schedule**: Branch creation - At any time (no schedule defined),
Automerge - At any time (no schedule defined).

🚦 **Automerge**: Disabled by config. Please merge this manually once you
are satisfied.

♻ **Rebasing**: Whenever PR is behind base branch, or you tick the
rebase/retry checkbox.

🔕 **Ignore**: Close this PR and you won't be reminded about this update
again.

---

- [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check
this box

---

This PR was generated by [Mend Renovate](https://mend.io/renovate/).
View the [repository job
log](https://developer.mend.io/github/IBM/appconfiguration-rust-sdk).

<!--renovate-debug:eyJjcmVhdGVkSW5WZXIiOiI0MS45Ny4xMCIsInVwZGF0ZWRJblZlciI6IjQxLjEzMC4xIiwidGFyZ2V0QnJhbmNoIjoibWFpbiIsImxhYmVscyI6W119-->

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Signed-off-by: Travis <rschoe@de.ibm.com>

jgsogo committed:
refact: Clarify `belong_to_segment` algorithm (#118)

* Introduces a `MatchesEntity` trait that is implemented for `Segment`
and `Rule` so we can isolate the logic that applies to each one of them
when matching an entity.

---------

Signed-off-by: Javier G. Sogo <jgsogo@gmail.com>
Signed-off-by: Rainer Schoenberger <rschoe@de.ibm.com>
Co-authored-by: Travis <rschoe@de.ibm.com>
Co-authored-by: Rainer Schoenberger <36071440+rainerschoe@users.noreply.github.com>
Signed-off-by: Travis <rschoe@de.ibm.com>

jgsogo committed:
test: Test IBMCloudTokenProvider renew network call (#97)

A new test for `IBMCloudTokenProvider` renew API call

Signed-off-by: Javier G. Sogo <jgsogo@gmail.com>
Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
Remove wait until config available

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
start to remove condvar from config

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
Fix tests

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
cleanup

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
cleanup + use new wait in tests

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
use wait in test + doc

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
cleanup

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
minimize diff

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
Merge remote-tracking branch 'github/main' into wait_for_init_config

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
remove debug printout

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
Introduce waitable instead of mutex/condvar tuple

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
no ref

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe commented on src/network/live_configuration/update_thread_worker.rs:
```diff
@@ -93,7 +94,10 @@ impl<T: ServerClient> UpdateThreadWorker<T> {
+        let (current_mode_mutex, condition_variable) = &*self.current_mode;
+        let mut current_mode = current_mode_mutex.lock().unwrap();
+        *current_mode = CurrentMode::Defunct(result.clone());
+        condition_variable.notify_all();
```
this is now implemented

jgsogo commented on src/client/app_configuration_offline.rs:
```diff
@@ -55,4 +56,9 @@ impl ConfigurationProvider for AppConfigurationOffline {
+
+    fn wait_until_online(&self) {
+        error!("Waiting for AppConfigurationOffline to get online. This will never happen.");
+        std::thread::park(); // block forever
```
maybe `panic!`? Not sure if I prefer to be blocked forever, or crash dramatically ASAP 😅

jgsogo commented on src/models/configuration.rs:
```diff
@@ -192,6 +193,11 @@ impl ConfigurationProvider for Configuration {
+
+    fn wait_until_online(&self) {
+        error!("Waiting for Configuration to get online. This will never happen.");
+        std::thread::park(); // block forever
```
maybe `panic!`? Not sure if I prefer to be blocked forever, or crash dramatically ASAP 😅

jgsogo commented on src/network/live_configuration/update_thread_worker.rs:
```diff
@@ -103,14 +105,24 @@ impl<T: ServerClient> UpdateThreadWorker<T> {
+
+                self.current_mode
+                    .set(CurrentMode::Online)
+                    .map_err(|_| Error::CannotAcquireLock)?;
```
We already provide a `Into` implementation from the `PoisonError` to our `Error::CannotAcquireLock`? We don't need this `map_err` here


```suggestion
                self.current_mode
                    .set(CurrentMode::Online)?;
```

jgsogo commented on src/network/live_configuration/update_thread_worker.rs:
```diff
@@ -103,14 +105,24 @@ impl<T: ServerClient> UpdateThreadWorker<T> {
+                    .set(CurrentMode::Offline(
+                        CurrentModeOfflineReason::FailedToGetNewConfiguration,
+                    ))
+                    .map_err(|_| Error::CannotAcquireLock)?;
```
```suggestion
                self.current_mode
                    .set(CurrentMode::Offline(
                        CurrentModeOfflineReason::FailedToGetNewConfiguration,
                    ))?;
```

jgsogo commented on src/network/live_configuration/update_thread_worker.rs:
```diff
@@ -141,8 +153,11 @@ impl<T: ServerClient> UpdateThreadWorker<T> {
+                        .set(CurrentMode::Offline(
+                            CurrentModeOfflineReason::WebsocketClosed,
+                        ))
+                        .map_err(|_| Error::CannotAcquireLock)?;
```
```suggestion
                    self.current_mode
                        .set(CurrentMode::Offline(
                            CurrentModeOfflineReason::WebsocketClosed,
                        ))?;
```

jgsogo commented on src/network/live_configuration/update_thread_worker.rs:
```diff
@@ -151,8 +166,11 @@ impl<T: ServerClient> UpdateThreadWorker<T> {
+                    .set(CurrentMode::Offline(
+                        CurrentModeOfflineReason::WebsocketError,
+                    ))
+                    .map_err(|_| Error::CannotAcquireLock)?;
```
```suggestion
                self.current_mode
                    .set(CurrentMode::Offline(
                        CurrentModeOfflineReason::WebsocketError,
                    ))?;
```

jgsogo reviewed (commented):
Looks great! 

 * I think we can remove some `map_err`
 * Block the thread or panic?

rainerschoe committed:
Apply suggestion from @jgsogo

Co-authored-by: Javier G. Sogo <jgsogo@gmail.com>
Signed-off-by: Rainer Schoenberger <36071440+rainerschoe@users.noreply.github.com>

rainerschoe committed:
Update src/network/live_configuration/update_thread_worker.rs

Co-authored-by: Javier G. Sogo <jgsogo@gmail.com>
Signed-off-by: Rainer Schoenberger <36071440+rainerschoe@users.noreply.github.com>

rainerschoe committed:
Update src/network/live_configuration/update_thread_worker.rs

Co-authored-by: Javier G. Sogo <jgsogo@gmail.com>
Signed-off-by: Rainer Schoenberger <36071440+rainerschoe@users.noreply.github.com>

rainerschoe committed:
Update src/network/live_configuration/update_thread_worker.rs

Co-authored-by: Javier G. Sogo <jgsogo@gmail.com>
Signed-off-by: Rainer Schoenberger <36071440+rainerschoe@users.noreply.github.com>

rainerschoe committed:
format

Signed-off-by: Travis <rschoe@de.ibm.com>

rainerschoe committed:
panic instead of block thread

Signed-off-by: Travis <rschoe@de.ibm.com>

jgsogo reviewed (approved):
Really nice! Thanks!
